syntax = "proto3";

package glossary;
// CRUD операции
service GlossaryService {
  rpc Home(Empty) returns (HomeResponse);
  rpc GetAll(Empty) returns (GetAllResponse);
  rpc GetTerm(GetTermRequest) returns (TermResponse);
  rpc SearchTerm(SearchTermRequest) returns (SearchTermResponse);
  rpc CreateTerm(CreateTermRequest) returns (TermResponse);
  rpc UpdateTerm(UpdateTermRequest) returns (TermResponse);
  rpc DeleteTerm(DeleteTermRequest) returns (DeleteResponse);
}

message Empty {}

message HomeResponse {
  string message = 1;
}

message GetTermRequest {
  int32 id = 1;
}

message SearchTermRequest {
  string keyword = 1;
}

message CreateTermRequest {
  string term = 1;
  optional string description = 2;
}

message UpdateTermRequest {
  int32 id = 1;
  optional string term = 2;
  optional string description = 3;
}

message DeleteTermRequest {
  int32 id = 1;
}

message Term {
  int32 id = 1;
  string term = 2;
  optional string description = 3;
}

message TermResponse {
  int32 id = 1;
  string term = 2;
  optional string description = 3;
  optional string error = 4;
}

message GetAllResponse {
  map<int32, Term> terms = 1;
}

message SearchTermResponse {
  map<int32, Term> results = 1;
}

message DeleteResponse {
  string message = 1;
  optional string error = 2;
}